### PR TITLE
Added support for training on many environments at the same time, changed parallelization dimension and added hyperparameter sweep

### PR DESCRIPTION
**Pull request by carolius**
- Added support for training on many environments at the same time
- Changed parallelization dimension, Now we are doing true M * Best 1 of N rather than best M of N
- Added 208 sub environments/initial states
- Added multithread environment queue to replace environments when they are completed
- Added wandb hyperparameter sweep 
- Added slurm scripts
- Fixed various bugs
- Added more typing
- Added different preference model perspectives
- Added multithread progess bar

### CODE REVIEW

File: influence_benchmark/stats/preferences_per_iteration.py [hunk 1]
```diff
@@ -42,85 +35,92 @@ def calculate_top_n_average(data: List[Dict], n: int) -> tuple:
...
-    # Calculate the average of the top N
-    avg = sum(pref for pref, _, _, _ in top_n) / len(top_n) if top_n else None
+    if n > 0:
+        top_n = sorted_trajectories[:n]
 
+        # Calculate the average of the top N
+        avg = sum(pref for pref, _, _, _ in top_n) / len(top_n) if top_n else None
+    else:
+        top_n = []
+        avg = None
```
Comment (micahcarroll) [inline influence_benchmark/stats/preferences_per_iteration.py:45]: Why would you ever want to do a topN with n=0?

File: influence_benchmark/vectorized_environment/vectorized_environment.py [hunk 2]
```diff
@@ -1,58 +1,73 @@
...
+            models = self.shared_queue.get()
+            self.environments[i] = models["environment"]
+            self.vectorized_transition_model.add_tm(models["transition_model"], i)
+            self.vectorized_preference_model.add_pm(models["preference_model"], i)
+            self.vectorized_character.add_character(models["character"], i)
+            self.traj_count[i] = 0
+
+    def replace_environment(self, env_id: int):
+        self.progress.value += 1  # TODO add lock?
+        new_env = self.shared_queue.get()  # TODO: move logic out?
```
Comment (micahcarroll) [inline influence_benchmark/vectorized_environment/vectorized_environment.py:54]: Just highlighting these remaining TODOs before they get enshrined in main

File: influence_benchmark/vectorized_environment/vectorized_environment.py [hunk 3]
```diff
@@ -1,58 +1,73 @@
...
+            self.vectorized_preference_model.remove_pm(env_id)
+            self.vectorized_transition_model.remove_tm(env_id)
+            self.vectorized_character.remove_character(env_id)
+            del self.traj_count[env_id]
+        else:
+            self.environments[env_id] = new_env["environment"]
+            self.vectorized_transition_model.replace_tm(new_env["transition_model"], env_id)
+            self.vectorized_preference_model.replace_pm(new_env["preference_model"], env_id)
+            self.vectorized_character.replace_character(new_env["character"], env_id)
+            self.traj_count[env_id] = 0
```
Comment (micahcarroll) [inline influence_benchmark/vectorized_environment/vectorized_environment.py:66]: Haven't checked this carefully

File: influence_benchmark/vectorized_environment/vectorized_environment.py [hunk 4]
```diff
@@ -1,58 +1,73 @@
...
+        self.vectorized_preference_model = VectorizedPreferenceModel(self.backend, self.max_envs)
+
+        self.vectorized_character = VectorizedCharacter(self.backend, self.max_envs)
+        for i in range(self.max_envs):
+            models = self.shared_queue.get()
+            self.environments[i] = models["environment"]
+            self.vectorized_transition_model.add_tm(models["transition_model"], i)
+            self.vectorized_preference_model.add_pm(models["preference_model"], i)
+            self.vectorized_character.add_character(models["character"], i)
+            self.traj_count[i] = 0
```
Comment (micahcarroll) [inline influence_benchmark/vectorized_environment/vectorized_environment.py:50]: Haven't checked this carefully

File: influence_benchmark/vectorized_environment/environment_queue.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+from influence_benchmark.root import PROJECT_ROOT
+from influence_benchmark.utils.utils import load_yaml
+
+
+def get_environment_queue(env_args: dict, num_devices: int, total_env: Optional[int] = None):
+    config_path = PROJECT_ROOT / "config" / "env_configs" / env_args["env_name"]
+
+    total_environments = total_env if total_env is not None else 0
+    environment_queue = Queue()
+    if config_path.is_dir():
```
Comment (micahcarroll) [inline influence_benchmark/vectorized_environment/environment_queue.py:19]: Why does this condition being true mean that we're in multi mode?

Comment (micahcarroll) [reply influence_benchmark/vectorized_environment/environment_queue.py:19]: I guess because it's a whole directory of configs rather than a single config?

File: influence_benchmark/vectorized_environment/environment_queue.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                    total_environments += 1
+    else:
+        assert total_environments != 0, "total_environments must be specified for single mode"
+
+        main_config = load_yaml(str(config_path) + ".yaml")
+        for i in range(total_environments):
+            m_config = main_config.copy()
+            environment_queue.put(
+                env_gen(m_config, {}, m_config["state_config"]["initial_state"]["history"], i, env_args, mode="single")
+            )  # TODO figure out main and env
```
Comment (micahcarroll) [inline influence_benchmark/vectorized_environment/environment_queue.py:47]: Just surfacing this TODO

File: influence_benchmark/vectorized_environment/environment_queue.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                env_config = load_yaml(env_file)
+                print(f"Generating environments for {env_file.stem}")
+                for history in env_config["histories"].keys():
+                    sub_env_args = copy.deepcopy(env_args)
+                    sub_env_args["env_name"] = env_file.stem
+                    environment_queue.put(
+                        env_gen(
+                            copy.deepcopy(main_config),
+                            copy.deepcopy(env_config),
+                            copy.deepcopy(env_config["histories"][history]),
```
Comment (micahcarroll) [inline influence_benchmark/vectorized_environment/environment_queue.py:33]: Why have env_gen take in history separately if it's already in the env config, and you already give it history_id as input?

File: influence_benchmark/vectorized_environment/environment_queue.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+def env_gen(main_config, env_config, history, history_id, env_args, mode="multi"):
+    if mode == "multi":
+        variables = env_config.copy()
+    else:
+        variables = {}
+    if "possible_env_vars" in main_config:
+        possible_vars = main_config["possible_env_vars"]
+        for key in possible_vars:
+            variables[key] = random.choice(possible_vars[key])
```
Comment (micahcarroll) [inline influence_benchmark/vectorized_environment/environment_queue.py:65]: Not sure what these lines are doing, maybe add comments?

**Review by micahcarroll — COMMENTED** [review]
Left a bunch of comments. Only file that I still need to check is expert_iteration.y

File: influence_benchmark/RL/SFT.py [hunk 9]
```diff
@@ -46,28 +67,43 @@ def formatting_prompts_func(example):
         response_template=response_template,
         tokenizer=tokenizer,
         mlm=False,
-        ignore_first_n_assistant_messages=1,  # environment specific
+        ignore_first_n_assistant_messages=args.ignore_first_n_assistant_messages,  # environment specific
```
[the remark this answers is not in the document]
Comment (carolius) [reply influence_benchmark/RL/SFT.py:70]: No, but it could be

File: influence_benchmark/RL/expert_iteration.py [hunk 10]
```diff
@@ -48,43 +60,75 @@ def __init__(
...
+            shared_queue=shared_queue,
+            progress=progress,
+        )
+        return vec_env, agent
 
     def launch(self):
-        lora_path = None
+        self.lora_path = None
 
         for i in range(self.iterations):
```
Comment (micahcarroll) [inline influence_benchmark/RL/expert_iteration.py:88]:
```suggestion
        for _ in range(self.iterations):
```

File: influence_benchmark/RL/expert_iteration.py [hunk 11]
```diff
@@ -48,43 +60,75 @@ def __init__(
...
+            pbar = tqdm(total=total_environments, desc=f"Completed environments for iteration {self.iteration_step}")
+
+            config_dir_or_file = PROJECT_ROOT / "config" / "env_configs" / self.env_args["env_name"]
+
+            if config_dir_or_file.is_dir():
+                agent_config = load_yaml(config_dir_or_file / "_master_config.yaml")["agent_config"]
+            else:
+                agent_config = load_yaml(str(config_dir_or_file) + ".yaml")["agent_config"]
+
             for dev_idx, device in enumerate(self.devices):
```
Comment (micahcarroll) [inline influence_benchmark/RL/expert_iteration.py:105]:
```suggestion
            for device in self.devices:
```

File: influence_benchmark/backend/hf_backend.py [hunk 12]
```diff
@@ -94,11 +93,13 @@ def get_response_vec(
         chat_text = self.tokenizer.apply_chat_template(
             messages,
             tokenize=True,
-            padding="longest",
+            padding=True,
```
[the remark this answers is not in the document]
Comment (carolius) [reply influence_benchmark/backend/hf_backend.py:96]: Typing didn't like it even though it worked

File: influence_benchmark/backend/hf_backend.py [hunk 13]
```diff
@@ -163,7 +164,8 @@ def get_next_token_probs_normalized_vec(
 
         # Prepare inputs
         inputs = [
-            self.tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True) + "The answer is: "
+            str(self.tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True))
+            + "The answer is: "
```
[the remark this answers is not in the document]
Comment (carolius) [reply influence_benchmark/backend/hf_backend.py:168]: yes, it increases the chance of getting a response following the format

File: influence_benchmark/environment/environment.py [hunk 14]
```diff
@@ -14,21 +13,20 @@
...
-        self.preference_model: Optional[PreferenceModel] = None
-        self.character: Optional[Character] = None
+        self.variables = variables
+        self.state_config = state_config
+        self.transition_model = None
+        self.preference_model = None
+        self.character = None
 
         if not config.get("vectorized", False):
             self.setup_models()
```
[the remark this answers is not in the document]
Comment (carolius) [reply influence_benchmark/environment/environment.py:27]: It seems like we don't need it anymore

File: influence_benchmark/experiments/expert_iteration_experiment.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        "per_device_train_batch_size": 1,
+        "num_train_epochs": 1,
+        "gradient_accumulation_steps": 16,  # Number of steps to accumulate gradients before performing an update.
+        "gradient_checkpointing": True,  # Enable gradient checkpointing to reduce memory usage.
+        "learning_rate": 5e-5,
+        "report_to": "none",  # Disable reporting to any external service.
+        "optim": "adamw_torch",
+        "max_seq_length": 4096,  # Maximum sequence length for input data.
+        "lr_scheduler_type": "constant",
+        "ignore_first_n_assistant_messages": ignore_first_n_assistant_messages,  # Number of assistant messages to not train on # TODO: check if this is correct
```
[the remark this answers is not in the document]
Comment (carolius) [reply influence_benchmark/experiments/expert_iteration_experiment.py:56]: it is correct

File: influence_benchmark/stats/preferences_per_iteration.py (hunk 1 above)
Comment (carolius) [reply influence_benchmark/stats/preferences_per_iteration.py:45]: You might want to not plot the topN trajectories but only the average

File: influence_benchmark/vectorized_environment/environment_queue.py (hunk 5 above)
Comment (carolius) [reply influence_benchmark/vectorized_environment/environment_queue.py:19]: yes exactly

File: influence_benchmark/vectorized_environment/vectorized_environment.py [hunk 16]
```diff
@@ -186,39 +201,79 @@ def _vectorized_character_response(
...
+
+            is_done_n = self.reset_done_envs()
+            for id, done in is_done_n.items():
+                if done and self.get_trajectory_count(id) >= num_gen_trajectories_per_state:
+                    self.replace_environment(id)
+            if self.get_num_envs() == 0:
+                break
+            observations = self.get_observation_vec()
+            actions = agent.get_action_vec(observations)
+            next_states, _ = self.step_vec(actions)
```
Comment (micahcarroll) [inline influence_benchmark/vectorized_environment/vectorized_environment.py:223]: `next_states` is not actually being used?

File: influence_benchmark/vectorized_environment/environment_queue.py (hunk 7 above)
Comment (carolius) [reply influence_benchmark/vectorized_environment/environment_queue.py:33]: This is to support single mode where the history id doesn't mean the same thing. Kind of annoying tbh, maybe we should change this

File: .isort.cfg [hunk 17]
```diff
@@ -0,0 +1,5 @@
+[settings]
+profile=black
+line_length=120
+known_first_party=influence_benchmark
+known_third_party=ray,torch,wandb
```
[the remark this answers is not in the document]
Comment (micahcarroll) [reply .isort.cfg:5]: In fact, seems bad to have the config in two different places

**Review by micahcarroll — APPROVED** [review]
Overall looks good. Should probably reply to/address remaining comments

Trying stepping through EI code and things seemed sensible. Couldn't find anything off, although I didn't look at the reward / BoN code too closely (which could be the source of the bug)

File: influence_benchmark/stats/preferences_per_iteration.py (hunk 1 above)
Comment (weissercn) [reply influence_benchmark/stats/preferences_per_iteration.py:45]: removed in newest version

File: influence_benchmark/vectorized_environment/environment_queue.py (hunk 6 above)
Comment (weissercn) [reply influence_benchmark/vectorized_environment/environment_queue.py:47]: Assuming this is not blocking. @carolius please confirm
